[FEATURE] Added Catalog Product Scope Hints

lewisvoncken opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios,
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
When there is a value set (when `Use Default` has been disabled) on website or storeview level there will be shown a notice in the Product Form

### Manual testing scenarios
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Set up a Multi Website
2. Create a Product through the Admin
3. Change the Scope to a storeview
4. Add a Storeview or website Specific Value
5. Change the Scope to the Default Scope
6. Now you can see that there is a Storeview or Website Specific value for the changed attribute
7. Click on the link and then it will open the Product Edit Page on that specific Storeview (when it is on Website scope it will open the default store of that website because it is not possible to change scope to website)

![image](https://user-images.githubusercontent.com/6040343/46596266-370cc600-cadc-11e8-873c-73fc574f6aec.png)


### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

lewisvoncken committed:
[FEATURE] Added Catalor Product Scope Hints

 - When there is a value set (when `Use Default` has been disabled) on website or storeview level there will be shown a notice in the Product Form

sidolov commented on app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Eav.php:
```diff
@@ -614,6 +624,149 @@ private function isProductExists()
+     */
+    private function formatOverriddenScopes(array $overridden)
+    {
+        $formatted = '<div class="overridden-hint-wrapper">' .
```
Try to render HTML in the view layer (html file for the field). In the modifier, we are allowed to return only raw data

lewisvoncken replied:
@sidolov 
Thanks for the feedback this has been resolved.

sidolov commented on app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Eav.php:
```diff
@@ -260,7 +268,8 @@ public function __construct(
         ScopeConfigInterface $scopeConfig = null,
-        AttributeCollectionFactory $attributeCollectionFactory = null
+        AttributeCollectionFactory $attributeCollectionFactory = null,
+        Url $urlBuilder
```
Please, inject new dependencies according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)

lewisvoncken replied:
@sidolov 
I have just resolved this. Thanks for the feedback

lewisvoncken committed:
[TASK] Inject dependencies according to Backward Compatible Development Guide

lewisvoncken committed:
[TASK] Render HTML in the view layer

lewisvoncken committed:
[TASK] Updated the according to Codacy Code Review

Solved:
The method getOverriddenLevels() has a Cyclomatic Complexity of 10. The configured cyclomatic complexity threshold is 10.

lewisvoncken commented:
@sidolov the requested changes have been made.

sidolov commented on app/code/Magento/Ui/view/base/web/templates/form/field.html:
```diff
@@ -40,6 +40,18 @@
 
         <div class="admin__additional-info" if="$data.additionalInfo" html="$data.additionalInfo"></div>
 
+        <div class="admin__scope-hints" if="$data.scopeHints">
```
Please, create the new custom template especially for the product form, it's not a good idea to define `scopeHints ` for all form fields

lewisvoncken replied:
@sidolov 
It won't be shown when there are no scopeHints and I think that there has to be changed to much to only use a separate template but correct me if I am wrong. Besides that it is also easy extendable for the Category Edit

sivaschenko replied:
@lewisvoncken the problem is that your changes are introduced to `app/code/Magento/Ui/view/base/web/templates/form/field.html` that is used for all fields, not only product, but the PHP implementation is related only to catalog product. Please create a template that will be used only for catalog product form.

sidolov commented on app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/Eav.php:
```diff
@@ -614,6 +625,132 @@ private function isProductExists()
+        }
+        $contextScopeId = $this->locator->getStore()->getId();
+
+        switch ($attribute->getScope()) {
```
please, use if statement for current logic, it will be more readable

lewisvoncken commented:
Hi @sidolov 
I will try to provide fixes asap.

sivaschenko commented:
@lewisvoncken closing this pull request due to inactivity. Feel free to reopen if you'd like to continue the progress on it.
